src/create_state.rs

```diff
@@ -59,8 +60,31 @@ pub fn create_state() -> Result<State, Error> {
         only: tor_only,
     });
 
-    Ok(State {
-        bind: (config.bind_address, config.bind_port).into(),
+    // not const/static because from() is not const
+    let default_bind_address = std::net::IpAddr::from([127, 0, 0, 1]);
+    // for consistency
+    let default_bind_port = 8331;
+
+    let bind_addr = match (config.bind_address, config.bind_port, config.bind_systemd_socket_name) {
+        (Some(addr), Some(port), None) => (addr, port).into(),
+        (None, Some(port), None) => (default_bind_address, port).into(),
+        (Some(addr), None, None) => (addr, default_bind_port).into(),
+        (None, None, None) => (default_bind_address, default_bind_port).into(),
+        (None, None, Some(socket_name)) => {
+            SocketAddr::from_systemd_name(socket_name)
+                .map_err(|error| {
+                    let new_error = anyhow::anyhow!("failed to parse systemd socket name: {}", error);
+                    slog::error!(logger, "failed to parse systemd socket name"; "error" => #error);
```

Comment:
This will result in a redundant error message, no?

Reply:
Also, shouldn't the caller be responsible for logging errors?

Reply:
The issue is `anyhow::Error` doesn't impl `std::error::Error` and I want to use the error logging feature of `slog`. I plan to replace `anyhow` with something better in the future.

Reply:
Didn't realize it doesn't. That seems like a wild oversight on behalf of the anyhow maintainers.

Reply:
Not an oversight, it's impossible to implement it because of coherence:

* `anyhow::Error` impls `From<T> where T: std::error::Error` intentionally, this is an important feature of anyhow
* `impl<T> From<T> for T` exists in `core`
* as a result `impl std::error::Error for anyhow::Error` would lead to duplicated impl of `From<anyhow::Error>`

A wrapper type would help, but I think removing `anyhow::Error` entirely is a cleaner approach.

Reply:
Ah. Similar to how failure has a [Compat](https://docs.rs/failure/0.1.8/failure/struct.Compat.html) type. 
I'm fine either way.